sbncode/SBNEventWeight/App/SystToolsEventWeight_module.cc

```diff
@@ -0,0 +1,238 @@
+  produces<std::vector<sbn::evwgh::EventWeightParameterSet>, art::InRun>();
+
+}
+
+
+void SystToolsEventWeight::produce(art::Event& e) {
+
+  auto mcwghvec = std::make_unique<std::vector<sbn::evwgh::EventWeightMap> >();
+  auto wghassns = std::make_unique<art::Assns<simb::MCTruth, sbn::evwgh::EventWeightMap> >();
+
+  art::PtrMaker<sbn::evwgh::EventWeightMap> makeWeightPtr(e);
+
+  std::vector<art::Ptr<simb::MCTruth> > mclist;
+  art::Handle<std::vector<simb::MCTruth>> mcTruthHandle;
+  if(!fGenieModuleLabel.empty()) e.getByLabel(fGenieModuleLabel, mcTruthHandle);
+  // Prooceed even with missing handle if we want to require the MCTruth to be
+  // found, so that an exception will be thrown explaining the problem.
+  if(mcTruthHandle.isValid() || !fAllowMissingTruth){
+
+    art::fill_ptr_vector(mclist, mcTruthHandle);
+
+    if(fDebugMode){
+      std::cout << "[SystToolsEventWeight::produce] mclist.size() = " << mclist.size() << "\n"
+                << "[SystToolsEventWeight::produce] fSystProviders.size() = " << fSystProviders.size() << std::endl;
+    }
+
+    for( auto &sp : fSystProviders ){
+
+      std::unique_ptr<systtools::EventResponse> syst_resp = sp->GetEventResponse(e);
+      if( !syst_resp ){
+        std::cout << "[ERROR]: Got nullptr systtools::EventResponse from provider "
+                  << sp->GetFullyQualifiedName();
+        throw std::exception();
+      }
+
+      //==== syst_resp->size() is (Number of MCTruth)
+      //==== Each index corresponds to each of MCTruth
+      int nMCTruthIndex(0);
+      if(fDebugMode) std::cout << "[SystToolsEventWeight::produce]   syst_resp.size() (= Number of MCTruth) of this SystProvider = " << syst_resp->size() << std::endl;
+
+      //==== Looping over syst_resp is identical to looping over MCTruth
+      for( systtools::EventResponse::iterator itResp = syst_resp->begin(); itResp != syst_resp->end(); ++itResp ){
+
+        systtools::event_unit_response_t resp = *itResp;
+        //==== resp.size() corresponds to number of knobs we altered;
+        //==== e.g., MaCCQE, MaCCRES, MvCCRE -> resp.size() = 3
+        if(fDebugMode){
+          std::cout << "[SystToolsEventWeight::produce]     sp->GetSystMetaData().size() (expected) = " << sp->GetSystMetaData().size() << "\n"
+                    << "[SystToolsEventWeight::produce]     resp.size() of this syst_resp (produced) = " << resp.size() << std::endl;
+        }
+        if(sp->GetSystMetaData().size()!=resp.size()){
+          std::cerr << "[SystToolsEventWeight::produce]     sp->GetFullyQualifiedName() = " << sp->GetFullyQualifiedName() << std::endl;
+          std::cerr << "[SystToolsEventWeight::produce]     We expect to have " << sp->GetSystMetaData().size() << " knobs for this SystProvider, but "
+                    << resp.size() << " are produced. "
+                    << "Probably this particular event is not relevant to this systematic variation." << std::endl;
+          throw std::exception();
```

Comment:
As above:
```suggestion
          throw cet::exception{ "SystToolsEventWeight" } 
            << "sp->GetFullyQualifiedName() = " << sp->GetFullyQualifiedName() << "\n"
            << "We expect to have " << sp->GetSystMetaData().size() << " knobs for this SystProvider, but "
                    << resp.size() << " are produced. "
                    << "Probably this particular event is not relevant to this systematic variation.\n";
```
The general idea is to use a pattern supported by the framework and to give the users the chance to decide how to configure the framework to cope with the errors.